### PR TITLE
Full implement group display names

### LINKED ISSUE #742: Add display name for groups
Group ids are being used and shown throughout the app, and cannot be changed.
Having a display name that can be changed (by admins?), and that are displayed when the group name must be visible to the user would be great.

### PR DESCRIPTION
**Pull request by nickvergessen**
- [x] Personal settings show group id instead of display names in list
- [x] Activities use group id instead of display names
- [x] Select2 in the WorkflowEngine shows the id on load, dropdown and selecting new values works
- [x] Select2 in the OC.Settings.setupGroupsSelect shows the id on load, dropdown and selecting new values works
- [x] User management: user group and subadmin selection is messed up
- [x] Tests

Need someone with better JS/select2 skills to solve this.

### Good patch for testing
```diff
diff --git a/lib/private/Group/Group.php b/lib/private/Group/Group.php
index 6795f0e2f5..24e8a7ad00 100644
--- a/lib/private/Group/Group.php
+++ b/lib/private/Group/Group.php
@@ -87,9 +87,9 @@ class Group implements IGroup {
 
        public function getDisplayName() {
                if (is_null($this->displayName)) {
-                       return $this->gid;
+                       return 'DisplayName{' . $this->gid . '}';
                }
-               return $this->displayName;
+               return 'DisplayName{' . $this->displayName . '}';
        }
 
        /**
```

Fixes #742

### CODE REVIEW

**Comment by nickvergessen** [comment]
@nextcloud/javascript who could help?

**Comment by MorrisJobke** [comment]
> @nextcloud/javascript who could help?

maybe @danxuliu or @skjnldsv

**Comment by skjnldsv** [comment]
Select2 issue is a big one!
Basically, since we store only the value of the select (aka groupid) and considering the fact that we only retrieve the groups data onclick on the select, the default text can't be anything else than the group id.
One solution would be to create the select as html with all of its options and apply select2 directly on it. I didn't test it, but it should work

**Comment by skjnldsv** [comment]
I'll change my view! I tried a different approach where I fetch the groups on page load only (no need to fetch them on each select2 click).
So I added on handlebar Views an ajax requests that get all the groups and cache them.
Then re-render the full select2. It works perfectly! :)

It quickly displays the id for a short moment, the time to requests the groups, then update itself with the proper displayName!

**Comment by skjnldsv** [comment]
I could use a hand on tests fixing for the last tests please @nickvergessen :)

**Comment by skjnldsv** [comment]
Test passes, failure unrelated :)

Please review

File: apps/workflowengine/js/admin.js [hunk 1]
```diff
@@ -161,6 +162,24 @@
...
 				}
+				var self = this;
+				$.ajax({
+					url: OC.generateUrl('settings/users/groups'),
+					dataType: 'json',
+					quietMillis: 100,
+				}).done(function(response) {
+					// add admin groups
+					$.each(response.data.adminGroups, function(id, group) {
+						self.groups.push({ id: group.id, displayname: group.name+'FIXME' });
```
Comment (nickvergessen) [inline apps/workflowengine/js/admin.js]: Still broken

File: apps/workflowengine/js/admin.js [hunk 2]
```diff
@@ -161,6 +162,24 @@
...
+					});
+					// add groups
+					$.each(response.data.groups, function(id, group) {
+						self.groups.push({ id: group.id, displayname: group.name+'FIXME' });
```
Comment (nickvergessen) [inline apps/workflowengine/js/admin.js]: Still broken

File: settings/js/settings.js [hunk 3]
```diff
@@ -73,12 +73,10 @@ OC.Settings = _.extend(OC.Settings, {
...
 				initSelection: function(element, callback) {
-					var selection =
-						_.map(($(element).val() || []).split('|').sort(),
-							function(groupName) {
+					var selection = _.map(($(element).val() || []).split('|').sort(), function(groupId) {
 						return {
-							id: groupName,
-							displayname: groupName
+							id: groupId,
+							displayname: groupId + 'FIXME' // FIXME
```
Comment (nickvergessen) [inline settings/js/settings.js]: Still broken (test with "exclude group from sharing")

File: apps/workflowengine/js/admin.js (hunk 1 above)
Comment (nickvergessen) [reply apps/workflowengine/js/admin.js]: okay, here it works, just need to remove the FIXME

File: apps/workflowengine/js/admin.js (hunk 2 above)
Comment (nickvergessen) [reply apps/workflowengine/js/admin.js]: okay, here it works, just need to remove the FIXME

File: apps/workflowengine/js/admin.js (hunk 1 above)
Comment (skjnldsv) [reply apps/workflowengine/js/admin.js]: Ahaha oups! :see_no_evil:

File: settings/js/settings.js (hunk 3 above)
Comment (skjnldsv) [reply settings/js/settings.js]: In file access control or workflow?

**Comment by nickvergessen** [comment]
- [ ] :x: Creating a new user and also adding them to a group (while creating the user) is broken, send data for groups is corrupted...
```
email |  
groups[ug-test][displayName] | DisplayName{ug-test}
password | 12346
username | test2e3e
```

File: settings/Controller/GroupsController.php [hunk 4]
```diff
@@ -108,13 +109,9 @@ public function create($id) {
...
 			);
 		}
-		if($this->groupManager->createGroup($id)) {
-			return new DataResponse(
-				array(
-					'groupname' => $id
-				),
-				Http::STATUS_CREATED
-			);
+		$group = $this->groupManager->createGroup($id);
```
Comment (blizzz) [inline settings/Controller/GroupsController.php]: there's not method to set the group name? not needed?

File: settings/js/settings.js (hunk 3 above)
Comment (skjnldsv) [reply settings/js/settings.js]: AH okay, I see. I did not see it was there :)

Comment (nickvergessen) [reply settings/js/settings.js]: in admin> sharing> 
> bildschirmfoto von 2018-03-05 13-07-20

File: settings/Controller/GroupsController.php (hunk 4 above)
Comment (nickvergessen) [reply settings/Controller/GroupsController.php]: No not needed atm

File: apps/workflowengine/js/usergroupmembershipplugin.js [hunk 5]
```diff
@@ -34,58 +34,36 @@
...
-						// add admin groups
-						$.each(response.data.adminGroups, function(id, group) {
-							results.push({ id: group.id });
+				data: groups,
+				initSelection: function (element, callback) {
+					var groupId = element.val();
+					if (groupId && groups.length > 0) {
+						callback({
+							id: groupId,
+							displayname: groups.find(group =>group.id === groupId).displayname
```
Comment (skjnldsv) [inline apps/workflowengine/js/usergroupmembershipplugin.js]: Hum, this is ES6, I'm not sure IE11 supports it :thinking:

**Comment by skjnldsv** [comment]
- [ ] search not working anymore

File: apps/workflowengine/js/usergroupmembershipplugin.js (hunk 5 above)
Comment (MorrisJobke) [reply apps/workflowengine/js/usergroupmembershipplugin.js]: We should not do this yet.

Comment (skjnldsv) [reply apps/workflowengine/js/usergroupmembershipplugin.js]: Fixed

**Comment by rullzer** [comment]
@skjnldsv https://github.com/nextcloud/server/pull/8255#issuecomment-369984358

> So I added on handlebar Views an ajax requests that get all the groups and cache them.

does this scale? Like even when you are on ldap and there are a few thousand groups?

**Comment by skjnldsv** [comment]
@rullzer I have no idea! The request was made anyway, I just did it before loading the select2 instead of after initialising it :)

**Comment by skjnldsv** [comment]
Failure related:
  ```
Scenario: Getting subadmin groups                               # /drone/src/github.com/nextcloud/server/build/integration/features/provisioning-v1.feature:252
    Given As an "admin"                                           # FeatureContext::asAn()
[Wed Mar  7 09:44:38 2018] 127.0.0.1:53098 [200]: /ocs/v2.php/cloud/users/brand-new-user
[Wed Mar  7 09:44:38 2018] 127.0.0.1:53108 [200]: /ocs/v2.php/cloud/users/brand-new-user
    And user "brand-new-user" exists                              # FeatureContext::assureUserExists()
[Wed Mar  7 09:44:38 2018] 127.0.0.1:53110 [200]: /ocs/v2.php/cloud/groups/new-group
[Wed Mar  7 09:44:38 2018] 127.0.0.1:53114 [200]: /ocs/v2.php/cloud/groups/new-group
    And group "new-group" exists                                  # FeatureContext::assureGroupExists()
[Wed Mar  7 09:44:38 2018] 127.0.0.1:53122 [200]: /ocs/v1.php/cloud/users/brand-new-user/subadmins
    When sending "GET" to "/cloud/users/brand-new-user/subadmins" # FeatureContext::sendingTo()
    Then subadmin groups returned are                             # FeatureContext::theSubadminGroupsShouldBe()
      | new-group |
      Failed asserting that two arrays are equal.
      --- Expected
      +++ Actual
      @@ @@
       Array (
      -    0 => 'new-group'
       )

```

**Comment by skjnldsv** [comment]
Failure unrelated: swift

**Comment by nickvergessen** [comment]
Backported the user facing stuff in https://github.com/nextcloud/server/pull/8779

**Review by blizzz — APPROVED** [review]
External storages shows group id, i'm okay if this gets adjusted in a follow up PR though.

**Review by rullzer — APPROVED** [review]
Lets do this and fix other occurences when we find them

**Pull request merged**